### PR TITLE
HIVE-26437: Dump unpartitioned tables in parallel.

### PR DESCRIPTION
**Pull request by atsaonerk**
Currently partitions of table is dump in parallel manner. But if table is not partitioned, it is dumped serially.
This change introduces parallelism at table level as well. A single thread pool which is currently being used
for partition level is also used for table level. The table level dump task is added to same thread pool.
The degree of parallelism depends upon config parameter REPL_PARTITIONS_DUMP_PARALLELISM whose defaul value is 100.
The new ExportService is introduced with this change which would be responsible for exporting table and partitions during repl dump. The ExportService is initialized and configured with thread pools by HiveServer2 service.
A new Hiveconfig variable ie "REPL_TABLE_DUMP_PARALLELISM is introduced to define the number of threads which would be created in thread pool.
The ExportService which is created as singleton instance would be used by ReplDumpTask.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 1]
```diff
@@ -952,6 +954,7 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
           }
         }
         for(String matchedDbName : Utils.matchesDb(hiveDb, work.dbNameOrPattern)) {
+          ExportService exportService = new ExportService(conf);
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:957]: Please move the ExportService object creation higher than a database level.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 2]
```diff
@@ -1254,6 +1273,7 @@ Long bootStrapDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive hiveDb)
             // Get the counter to store the snapshots created & deleted at source.
             replSnapshotCount = new SnapshotUtils.ReplSnapshotCount();
           }
+          ExportService exportService = new ExportService(conf);
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:1276]: Please move the ExportService object creation higher than a database level.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 3]
```diff
@@ -1290,6 +1310,21 @@ Long bootStrapDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive hiveDb)
             }
           }
 
+          if (exportService != null && exportService.isExportServiceRunning()) {
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:1313]: Please move the ExportService object creation higher than a database level.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 4]
```diff
@@ -978,6 +981,22 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
               LOG.debug(te.getMessage());
             }
           }
+
+          if (exportService != null && exportService.isExportServiceRunning()) {
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:985]: Please move the ExportService object creation higher than a database level.

**Review by pudidic — COMMENTED** [review]
Please move the ExportService object creation higher than a database level. It will make all databases share the same service, not creating one for each database.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 4 above)
Comment (atsaonerk) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:985]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 1 above)
Comment (atsaonerk) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:957]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 2 above)
Comment (atsaonerk) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:1276]: Done

**Comment by pudidic** [comment]
+1 Looks good to me. I tested on my laptop and the failing test passed. Please re-trigger the CI system to make sure it works.

**Comment by pudidic** [comment]
+1. LGTM. The failing test passed.

**Comment by rbalamohan** [comment]
LGTM. +1

**Review by lcspinter — APPROVED** [review]

**Pull request merged**